### DISCUSSION TITLE
How does NextRequest gets the ip from the request?

### DISCUSSION THREAD

**Discussion by iondrimba** [vercel/next.js#33435] [Help] [answered]
I'm trying to use the new middleware feature to log some information including the request IP. During my tests NextRequest.ip always returns undefined

I'm running a local express server and already tried to use https://www.npmjs.com/package/request-ip 
but the NextRequest object always returns ip = undefined

After installing request-ip middleware I now have request.clientIp, but when the request hits nextjs middleware request.ip is undefined

**Comment by joggienl** [comment, marked as answer]
Have you tried looking in the headers? Something like `req.headers['x-real-ip']` and/or `req.headers['x-forwarded-for']`.

Perhaps is this also helpful, an post on SO I found: https://stackoverflow.com/questions/68338838/how-to-get-the-ip-address-of-the-client-from-server-side-in-next-js-app

**Reply by wjm0913** [reply]
Hello, using the getInitialProps method, you cannot get the IP address when you first enter the page. You can only get the IP address when you refresh the browser? Do you know similar questions? Asking for advice

**Reply by paurush11** [reply]
Provided by Vercel
**request.ip**
Self-hosting
**request.headers.get('X-Forwarded-For') or request.headers.get('x-real-ip')**

From the docs --> https://nextjs.org/docs/app/api-reference/functions/next-request
